File: deploy_ftp.py
import os
import sys
import ftplib
from pathlib import Path

FTP_HOST = '115.68.168.243'
FTP_USER = 'nuriohtrade'
FTP_PASS = 'seungho0409#'
REMOTE_ROOT = '/public_html'

LOCAL_BASE = Path(__file__).resolve().parent
DIST_DIR = LOCAL_BASE / 'dashboard' / 'dist'
PHP_DIR = LOCAL_BASE / 'php'

def make_dirs(ftp, remote_path):
    parts = [p for p in remote_path.replace('\\', '/').split('/') if p]
    current = ""
    for part in parts:
        current += "/" + part
        try:
            ftp.cwd(current)
        except ftplib.error_perm:
            try:
                ftp.mkd(current)
                print(f"📁 Created remote dir: {current}")
            except Exception as e:
                pass

def upload_file(ftp, local_file, remote_file):
    remote_dir = '/'.join(remote_file.replace('\\', '/').split('/')[:-1])
    make_dirs(ftp, remote_dir)
    ftp.cwd(remote_dir)
    
    filename = Path(remote_file).name
    print(f"⬆️ Uploading: {filename} -> {remote_dir}")
    with open(local_file, 'rb') as f:
        ftp.storbinary(f'STOR {filename}', f)

def main():
    print("🚀 Connecting to FTP server...")
    ftp = ftplib.FTP()
    ftp.connect(FTP_HOST, 21, timeout=30)
    ftp.login(FTP_USER, FTP_PASS)
    ftp.encoding = 'utf-8'
    print(f"✅ Connected to {FTP_HOST} as {FTP_USER}")

    # 1. Upload .htaccess
    htaccess_content = """<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteBase /

    # 1. API 라우팅
    RewriteRule ^api/(.*)$ api/index.php [QSA,L]
    RewriteRule ^api$ api/index.php [QSA,L]

    # 2. 존재하는 파일/디렉토리 제공
    RewriteCond %{REQUEST_FILENAME} -f [OR]
    RewriteCond %{REQUEST_FILENAME} -d
    RewriteRule ^ - [L]

    # 3. SPA 프론트엔드 라우팅
    RewriteRule ^ index.html [L]
</IfModule>
"""
    htaccess_local = LOCAL_BASE / '.htaccess'
    with open(htaccess_local, 'w', encoding='utf-8') as f:
        f.write(htaccess_content)
    upload_file(ftp, htaccess_local, f'{REMOTE_ROOT}/.htaccess')

    # 2. Upload dist files (Frontend)
    print(f"\n📦 Uploading Frontend build ({DIST_DIR})...")
    for root, dirs, files in os.walk(DIST_DIR):
        for file in files:
            local_path = Path(root) / file
            rel_path = local_path.relative_to(DIST_DIR)
            remote_path = f"{REMOTE_ROOT}/{rel_path.as_posix()}"
            upload_file(ftp, local_path, remote_path)

    # 3. Upload PHP files (Backend)
    print(f"\n🐘 Uploading PHP Backend files...")
    # api/index.php
    api_index = PHP_DIR / 'api' / 'index.php'
    if api_index.exists():
        upload_file(ftp, api_index, f"{REMOTE_ROOT}/api/index.php")
        upload_file(ftp, api_index, f"{REMOTE_ROOT}/php/api/index.php")

    # config/database.php
    db_config = PHP_DIR / 'config' / 'database.php'
    if db_config.exists():
        upload_file(ftp, db_config, f"{REMOTE_ROOT}/config/database.php")
        upload_file(ftp, db_config, f"{REMOTE_ROOT}/php/config/database.php")

    # migrate.php
    migrate_file = PHP_DIR / 'migrate.php'
    if migrate_file.exists():
        upload_file(ftp, migrate_file, f"{REMOTE_ROOT}/migrate.php")

    ftp.quit()
    print("\n🎉 All files uploaded successfully to /public_html on hosting server!")

if __name__ == '__main__':
    main()

File: deploy_ftp.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const FTP_HOST = '115.68.168.243';
const FTP_USER = 'nuriohtrade';
const FTP_PASS = 'seungho0409#';
const REMOTE_ROOT = '/public_html';

const LOCAL_BASE = process.cwd();
const DIST_DIR = path.resolve(LOCAL_BASE, 'dashboard', 'dist');
const PHP_DIR = path.resolve(LOCAL_BASE, 'php');

function uploadFileWithCurl(localPath, remoteRelativePath) {
  const remoteUrl = `ftp://${encodeURIComponent(FTP_USER)}:${encodeURIComponent(FTP_PASS)}@${FTP_HOST}${REMOTE_ROOT}/${remoteRelativePath.replace(/\\/g, '/')}`;
  console.log(`⬆️ Uploading: ${remoteRelativePath}`);
  try {
    execSync(`curl -s --ftp-create-dirs -T "${localPath}" "${remoteUrl}"`, { stdio: 'inherit' });
  } catch (err) {
    console.error(`❌ Failed to upload ${localPath}:`, err.message);
  }
}

function uploadDirectory(dir, baseDir) {
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const fullPath = path.join(dir, file);
    const stat = fs.statSync(fullPath);
    if (stat.isDirectory()) {
      uploadDirectory(fullPath, baseDir);
    } else {
      const relPath = path.relative(baseDir, fullPath);
      uploadFileWithCurl(fullPath, relPath);
    }
  }
}

async function main() {
  console.log('🚀 Starting FTP Deployment to iwinv hosting server...');

  // 1. Upload .htaccess
  const htaccessContent = `<IfModule mod_headers.c>
    # HTML, JS, JSON 캐시 방지 (항상 최신 버전 보장)
    <FilesMatch "\\.(html|htm|json)$">
        Header set Cache-Control "max-age=0, no-cache, no-store, must-revalidate"
        Header set Pragma "no-cache"
        Header set Expires "Wed, 11 Jan 1984 05:00:00 GMT"
    </FilesMatch>
    <FilesMatch "^sw\\.js$">
        Header set Cache-Control "max-age=0, no-cache, no-store, must-revalidate"
    </FilesMatch>
</IfModule>

<IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteBase /

    # 1. API 라우팅
    RewriteRule ^api/(.*)$ api/index.php [QSA,L]
    RewriteRule ^api$ api/index.php [QSA,L]

    # 2. 존재하는 파일/디렉토리 제공
    RewriteCond %{REQUEST_FILENAME} -f [OR]
    RewriteCond %{REQUEST_FILENAME} -d
    RewriteRule ^ - [L]

    # 3. SPA 프론트엔드 라우팅
    RewriteRule ^ index.html [L]
</IfModule>
`;
  const htaccessLocal = path.join(LOCAL_BASE, '.htaccess');
  fs.writeFileSync(htaccessLocal, htaccessContent, 'utf-8');
  uploadFileWithCurl(htaccessLocal, '.htaccess');

  // 2. Upload Frontend Dist
  console.log('\n📦 Uploading Frontend Build (dashboard/dist)...');
  uploadDirectory(DIST_DIR, DIST_DIR);

  // 3. Upload PHP Backend
  console.log('\n🐘 Uploading PHP Backend files...');
  const apiIndex = path.join(PHP_DIR, 'api', 'index.php');
  if (fs.existsSync(apiIndex)) {
    uploadFileWithCurl(apiIndex, 'api/index.php');
    uploadFileWithCurl(apiIndex, 'php/api/index.php');
  }

  const dbConfig = path.join(PHP_DIR, 'config', 'database.php');
  if (fs.existsSync(dbConfig)) {
    uploadFileWithCurl(dbConfig, 'config/database.php');
    uploadFileWithCurl(dbConfig, 'php/config/database.php');
  }

  const migrateFile = path.join(PHP_DIR, 'migrate.php');
  if (fs.existsSync(migrateFile)) {
    uploadFileWithCurl(migrateFile, 'migrate.php');
  }

  console.log('\n🎉 Deployment to http://nuriohtrade.iwinv.net completed successfully!');
}

main();
